Userland/Libraries/LibCompress/Brotli.cpp

```diff
@@ -0,0 +1,905 @@
+            }
+        }
+    }
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_prefix_code(CanonicalCode& code, size_t alphabet_size)
+{
+    size_t hskip = TRY(m_input_stream.read_bits(2));
+
+    if (hskip == 1) {
+        TRY(read_simple_prefix_code(code, alphabet_size));
+    } else {
+        TRY(read_complex_prefix_code(code, alphabet_size, hskip));
+    }
+
+    return {};
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_simple_prefix_code(CanonicalCode& code, size_t alphabet_size)
+{
+    size_t number_of_symbols = 1 + TRY(m_input_stream.read_bits(2));
+
+    size_t symbol_size = 0;
+    while ((1u << symbol_size) < alphabet_size)
+        symbol_size++;
+
+    Vector<size_t> symbols;
+    for (size_t i = 0; i < number_of_symbols; i++) {
+        size_t symbol = TRY(m_input_stream.read_bits(symbol_size));
+        symbols.append(symbol);
+    }
+
+    if (number_of_symbols == 1) {
+        code.m_symbol_codes.append(0b1);
+        code.m_symbol_values = move(symbols);
+    } else if (number_of_symbols == 2) {
+        code.m_symbol_codes.extend({ 0b10, 0b11 });
+        if (symbols[0] > symbols[1])
+            swap(symbols[0], symbols[1]);
+        code.m_symbol_values = move(symbols);
+    } else if (number_of_symbols == 3) {
+        code.m_symbol_codes.extend({ 0b10, 0b110, 0b111 });
+        if (symbols[1] > symbols[2])
+            swap(symbols[1], symbols[2]);
+        code.m_symbol_values = move(symbols);
+    } else if (number_of_symbols == 4) {
+        bool tree_select = TRY(m_input_stream.read_bit());
+        if (tree_select) {
+            code.m_symbol_codes.extend({ 0b10, 0b110, 0b1110, 0b1111 });
+            if (symbols[2] > symbols[3])
+                swap(symbols[2], symbols[3]);
+            code.m_symbol_values = move(symbols);
+        } else {
+            code.m_symbol_codes.extend({ 0b100, 0b101, 0b110, 0b111 });
+            quick_sort(symbols);
+            code.m_symbol_values = move(symbols);
+        }
+    }
+
+    return {};
+}
+
+ErrorOr<void> BrotliDecompressionStream::read_complex_prefix_code(CanonicalCode& code, size_t alphabet_size, size_t hskip)
+{
+    // hskip should only be 0, 2 or 3
+    VERIFY(hskip != 1);
+    VERIFY(hskip <= 3);
+
+    // Read the prefix code_value that is used to encode the actual prefix code_value
+    size_t const symbol_mapping[18] = { 1, 2, 3, 4, 0, 5, 17, 6, 16, 7, 8, 9, 10, 11, 12, 13, 14, 15 };
+    size_t code_length[18] { 0 };
+    size_t code_length_counts[6] { 0 };
+
+    size_t sum = 0;
+    size_t number_of_non_zero_symbols = 0;
+    for (size_t i = hskip; i < 18; i++) {
+        size_t len = TRY(read_complex_prefix_code_length());
+        code_length[symbol_mapping[i]] = len;
+
+        if (len != 0) {
+            code_length_counts[len]++;
+            sum += (32 >> len);
+            number_of_non_zero_symbols++;
+        }
+
+        if (sum == 32)
+            break;
+        VERIFY(sum < 32);
```

Comment:
This can be reached with this test file:
[crash-8e13ebed311edf16682ac6004d77d07000bc2933.txt](https://github.com/SerenityOS/serenity/files/8421866/crash-8e13ebed311edf16682ac6004d77d07000bc2933.txt)